From 01e57e530e31078795eb43b9a4b0c86f2c08579c Mon Sep 17 00:00:00 2001
From: Zhen Ni <zhen.ni@easystack.cn>
Date: Wed, 21 Jan 2026 07:04:21 -0500
Subject: [PATCH 03772/12543] dmaengine: fsl-edma: Fix clk leak on
 alloc_chan_resources failure

[ Upstream commit b18cd8b210417f90537d914ffb96e390c85a7379 ]

When fsl_edma_alloc_chan_resources() fails after clk_prepare_enable(),
the error paths only free IRQs and destroy the TCD pool, but forget to
call clk_disable_unprepare(). This causes the channel clock to remain
enabled, leaking power and resources.

Fix it by disabling the channel clock in the error unwind path.

Fixes: d8d4355861d8 ("dmaengine: fsl-edma: add i.MX8ULP edma support")
Cc: stable@vger.kernel.org
Suggested-by: Frank Li <Frank.Li@nxp.com>
Signed-off-by: Zhen Ni <zhen.ni@easystack.cn>
Reviewed-by: Frank Li <Frank.Li@nxp.com>
Link: https://patch.msgid.link/20251014090522.827726-1-zhen.ni@easystack.cn
Signed-off-by: Vinod Koul <vkoul@kernel.org>
[ Different error handling scheme ]
Signed-off-by: Sasha Levin <sashal@kernel.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---
 drivers/dma/fsl-edma-common.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/dma/fsl-edma-common.c b/drivers/dma/fsl-edma-common.c
index 561e3a71a5dd..c018e3769e3c 100644
--- a/drivers/dma/fsl-edma-common.c
+++ b/drivers/dma/fsl-edma-common.c
@@ -850,6 +850,7 @@ int fsl_edma_alloc_chan_resources(struct dma_chan *chan)
 			if (fsl_edma_drvflags(fsl_chan) & FSL_EDMA_DRV_HAS_PD)
 				pm_runtime_put_sync_suspend(fsl_chan->pd_dev);
 
+			clk_disable_unprepare(fsl_chan->clk);
 			return ret;
 		}
 
-- 
2.53.0

